[Added] Robin Hood hashmap

EinfachAndy opened the pull request:
* used Robin Hood algorithm to handle conflicts, which has a better distribution of the probe sequence lengths (PSL)
 * see: https://cs.uwaterloo.ca/research/tr/1986/CS-86-14.pdf
 * Note:
    - The expected length of the longest PSL (and thus the expected runtime complexity of lookup, remove and insert) in a full table is Θ(ln n)
    - insert operations have now more memory swaps of entries, to create the distribution

EinfachAndy commented:
Hi,
This patch should improve the look up performance of the hashmap using a different conflict strategy. The idea is described in detail in the linked paper. I did not check the performance, that is still an open task. Do you have any benchmark tools or benchmark code to do this?

zyedidia commented:
Very cool, thanks! I'll take a look. I don't have any hashmap benchmarks, but maybe we could make a simple one that just does random operations (like the test).

EinfachAndy commented:
I will add some simple benchmarks, which can also be improved later. Furthermore I can separate the code in a way, that your table still exsits and the new code is moved to hashmap/robin.go. Then we can compare both and also against the golang std::map.

zyedidia commented:
That sounds great.

zyedidia commented:
Looks good, thanks! Do you think this is ready to merge?

EinfachAndy commented:
I will take a final look in the next days and give you a sign for a final review..

EinfachAndy commented:
Hi,
the patches are ready for a review. I created a first benchmark run and linked the html output.

[chart.tar.gz](https://github.com/zyedidia/generic/files/11465798/chart.tar.gz)

zyedidia commented on hashmap/robin.go:
```diff
@@ -0,0 +1,252 @@
+}
+
+// NewRobinMap constructs a new map with default settings.
+func NewRobinMap[K comparable, V any]() *RobinMap[K, V] {
```
Is it important that the key be comparable? It would be nice to be consistent between the normal map and the robin map by using an equals and hash callback instead of requiring the key to be comparable. Hopefully that will also allow deletion of the GetHasher function which uses unsafe.

EinfachAndy replied:
The reason why I made the key comparable is because than I can use the flat compare operator `==`
here https://github.com/zyedidia/generic/pull/41/files#diff-3e6a5b1df35ab9e622b2b5800349c79c52dfd9156768960701f5917f1038402eR85

The problem is that the golang compiler does not optimize out static callback functions. So a flat compare is significant faster than a function call.

zyedidia commented on generic.go:
```diff
@@ -130,3 +191,52 @@ func hash(u uint64) uint64 {
 }
+
+// GetHasher returns a default hasher function for different Key types.
+func GetHasher[Key any]() HashFn[Key] {
```
It would be great to implement this without Unsafe, is that possible? Or better, remove the need for this by using `EqualsFn` and `HashFn` like the existing map.

EinfachAndy replied:
I removed the function and the unsafe import

zyedidia commented on generic.go:
```diff
@@ -41,6 +58,44 @@ func Compare[T any](a, b T, less LessFn[T]) int {
 
+// NextPowerOf2 is a fast calculation implementation of 2^x.
+// see: https://stackoverflow.com/questions/466204/rounding-up-to-next-power-of-2
+// go:inline
```
Ideally we wouldn't have these `go:inline` comments. Are they really necessary for good performance?

EinfachAndy replied:
fixed

zyedidia commented on hashmap/robin.go:
```diff
@@ -0,0 +1,252 @@
+	}
+}
+
+// go:inline
```
Same comment as above about `go:inline`.

EinfachAndy replied:
fixed

zyedidia commented:
Thanks! Looks quite good overall, and thanks for the benchmarks! I've added some comments about things that can hopefully be addressed.

EinfachAndy commented:
Hi,
I modified and move the code into another [repository](https://github.com/EinfachAndy/hashmaps), where I also implement some other kind of hashmaps.
The benchmark code is moved [here](https://github.com/EinfachAndy/bench-hashmaps). I also added your implementation to the benchmark and some other common like [swiss](https://github.com/dolthub/swiss)

I decided it, because currently your public hashmap interface is not so comfortable to use. E.g.:
 * Put and Remove does not return the state if the key is new / was in
 * Each has no stop option
 * The g.EqualsFn[K] compare call has a performance overhead
 * The Constuctor signature is annoying to use in a generic factory [see](https://github.com/EinfachAndy/bench-hashmaps/blob/main/bench.go#L161)

Sry, I hope my decision is understandable.

Best regards
Andy

zyedidia commented:
Alright it’s up to you, sorry for being slow to respond about your updates. I’m not opposed to having a different interface than the basic hashmap that is slightly more amenable to good performance/other niceties, but I am in general opposed to the use of unsafe or `go:inline` in this package. If you prefer to keep more performant hashmaps in a different repository then that’s understandable.
